### PR TITLE
[BM-10] As a user, I should be able to list cart contents

### PR DESCRIPTION
**Pull request by szcrubyrootstrap**
**Feature**
[BM-10](https://trello.com/c/JGMgO3A7/13-bm-10-as-a-user-i-should-be-able-to-list-cart-contents)
----------------
**Description**
The `/api/v1/cart` endpoint has been created that allows listing the products added to the shopping cart by the user. 

The information provided by the endpoint is serialized with the `Blueprinter` gem included in this branch. Also the `rails_best_practices` gem has been added to improve the source code.

A restructuring of the endpoints that allow adding and removing `Products` from the `Cart` has been generated, leaving the following endpoints to carry out said actions.

```
POST /api/v1/cart/products         --> Creates (Adds just one) Product to Cart.
DELETE /api/v1/cart/products/:id   --> Deletes all units of a Product from the Cart.
PUT /api/v1/cart/products/:id      --> Updates the Cart Products using positive or negatives values for adding or removing.
```

----------------
**Notes**
The following Json is the response structure as example for the products added by user to the shopping cart.

```
{
    "id": 7,
    "products": [
        {
            "id": 41646,
            "cart_products": [
                {
                    "id": 54,
                    "quantity": 2,
                    "total_amount": 72986
                }
            ],
            "description": "Provident iusto aperiam assumenda velit esse. Voluptate reprehenderit voluptates eligendi occaecati tenetur. Ut vitae veritatis quas accusamus expedita odit quos qui. Cum sit quidem eius ea ab placeat.",
            "name": "NWE Auto Transmitter",
            "price": 36493
        }
    ],
    "status": "in_process",
    "total_items": 2,
    "total_price": 72986.0,
    "user": {
        "id": 1,
        "email": "foo@baz.com"
    }
}
```

### CODE REVIEW

File: app/controllers/api/v1/carts_controller.rb [hunk 1]
```diff
@@ -1,6 +1,13 @@
 module Api
   module V1
     class CartsController < Api::V1::ApiController
+      def collection
+        cart = current_user.carts.find_by(status: :in_process)
```
Comment (mrodriguez90) [inline app/controllers/api/v1/carts_controller.rb]: this query is a good candidate for a model scope

**Review by mrodriguez90 — APPROVED** [review]
LGTM

File: app/controllers/api/v1/carts_controller.rb [hunk 2]
```diff
@@ -1,6 +1,13 @@
...
+        raise Api::CartNotFoundError, { user_id: current_user.id } unless cart
```
Comment (afmicc) [inline app/controllers/api/v1/carts_controller.rb]: what is the reason behind a custom exception for a not found error? i think it's not necessary at this point, the active record one should be enough

Comment (afmicc) [inline app/controllers/api/v1/carts_controller.rb]: i don't think it's a good idea to include the user id in the error message. In some context, it could be sensitive information that we don't want to share with clients.

File: app/serializer/cart_serializer.rb [hunk 3]
```diff
@@ -0,0 +1,12 @@
+class CartSerializer < Blueprinter::Base
```
Comment (afmicc) [inline app/serializer/cart_serializer.rb:1]: it looks so simple to use. Great choice!

File: config/routes.rb [hunk 4]
```diff
@@ -4,15 +4,11 @@
...
-      resource :carts, only: :show do
-        collection do
-          put :add_product
-          put :remove_product
-        end
-      end
+    namespace :v1, defaults: { format: :json } do
+      put :add_product, to: 'carts#add_product', path: 'carts/add_product'
+      put :remove_product, to: 'carts#remove_product', path: 'carts/remove_product'
+      get :collection, to: 'carts#collection', path: 'carts/collection'
```
Comment (afmicc) [inline config/routes.rb]: what is the reason behind this change? i prefer the previous notation

File: app/controllers/api/v1/carts_controller.rb [hunk 5]
```diff
@@ -1,6 +1,13 @@
 module Api
   module V1
     class CartsController < Api::V1::ApiController
+      def collection
```
Comment (JulianPasquale) [inline app/controllers/api/v1/carts_controller.rb]: We should follow the rest convention here

File: spec/requests/api/v1/carts/remove_product_spec.rb [hunk 6]
```diff
@@ -60,5 +56,13 @@
 
       it_behaves_like 'response bad_request'
     end
+
+    context 'when user not logged in' do
+      let(:headers) { nil }
+      let(:quantity) { 1 }
+      let(:product) { create(:product) }
+
+      it_behaves_like 'user is not logged in'
```
Comment (JulianPasquale) [inline spec/requests/api/v1/carts/remove_product_spec.rb]: Personally I don't use shared examples too much. Sometimes they make tests less descriptive and are not so useful for documentation.

I'd suggest to rename you shared_examples 
```suggestion
      it_behaves_like 'an unauthorised request'
```

https://www.betterspecs.org/#shared

File: spec/requests/api/v1/carts/collection_spec.rb [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+
+    it 'returns the correct response keys' do
+      expect(json.keys).to contain_exactly('id', 'products', 'status', 'total_items',
+                                           'total_price', 'user')
+    end
+
+    it 'returns the schema information specified in the serializer' do
+      data_serialized = CartSerializer.render_as_json(cart, view: :with_product_resume)
+
+      expect(json).to eq(data_serialized)
```
Comment (afmicc) [inline spec/requests/api/v1/carts/collection_spec.rb]: maybe you could explore the usage of the [json_expectations](https://github.com/waterlink/rspec-json_expectations) or any other similar gem that helps to validate the returned json

File: spec/requests/api/v1/carts/remove_product_spec.rb (hunk 6 above)
Comment (szcrubyrootstrap) [reply spec/requests/api/v1/carts/remove_product_spec.rb]: updated.

File: app/controllers/api/v1/carts_controller.rb (hunk 2 above)
Comment (szcrubyrootstrap) [reply app/controllers/api/v1/carts_controller.rb]: the custom error was removed.

Comment (szcrubyrootstrap) [reply app/controllers/api/v1/carts_controller.rb]: It's correct. Updated

File: config/routes.rb (hunk 4 above)
Comment (szcrubyrootstrap) [reply config/routes.rb]: It was changed to consider only available endpoints. The structure was re-established in the last commit 😄

File: spec/requests/api/v1/carts/collection_spec.rb (hunk 7 above)
Comment (szcrubyrootstrap) [reply spec/requests/api/v1/carts/collection_spec.rb]: Gem added and used for this endpoint verification.

File: app/controllers/api/v1/products_controller.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+      def create
+        cart_item_management.create_product_in_cart
+
+        render json: response_data, status: :created
+      end
+
+      def destroy
+        cart_item_management.remove_product_from_cart
+
+        render json: response_data, status: :ok
```
Comment (afmicc) [inline app/controllers/api/v1/products_controller.rb]: what data do you return here? maybe you can return a `head :no_content` here

File: app/services/cart_items_management_service.rb [hunk 9]
```diff
@@ -1,30 +1,39 @@
...
-  def initialize(cart, product, quantity)
+  def initialize(cart, product, quantity = 0)
     @cart = cart
     @product = product
     @quantity = quantity
+    @cart_product = CartProduct.find_by(cart: cart, product: product)
+  end
 
-    validate_quantity!
+  def create_product_in_cart
```
Comment (afmicc) [inline app/services/cart_items_management_service.rb]: regarding the name of the service is giving us the context that we are handling the cart contents, i would keep the names a bit shorter. wdyt?
```suggestion
  def create_product!
```
also, add the bang too

File: config/environments/development.rb [hunk 10]
```diff
@@ -6,7 +6,7 @@
     Bullet.alert         = true
     Bullet.bullet_logger = true
     Bullet.console       = true
-  # Bullet.growl         = true
+    # Bullet.growl         = true
```
Comment (afmicc) [inline config/environments/development.rb:9]: maybe you could just delete it

File: spec/requests/api/v1/carts/show_spec.rb [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+                            total_amount: total_amount)
+    end
+    let(:quantity) { 2 }
+    let(:total_amount) { product.price * quantity }
+
+    before { subject }
+
+    it_behaves_like 'a successful request'
+
+    it 'returns the schema information specified in the serializer' do
```
Comment (afmicc) [inline spec/requests/api/v1/carts/show_spec.rb]:
```suggestion
    it 'returns cart info' do
```
it'd be enough

**Review by afmicc — APPROVED** [review]

File: app/services/cart_items_management_service.rb (hunk 9 above)
Comment (szcrubyrootstrap) [reply app/services/cart_items_management_service.rb]: updated.

File: spec/requests/api/v1/carts/show_spec.rb (hunk 11 above)
Comment (szcrubyrootstrap) [reply spec/requests/api/v1/carts/show_spec.rb]: Updated.

File: app/controllers/api/v1/products_controller.rb (hunk 8 above)
Comment (szcrubyrootstrap) [reply app/controllers/api/v1/products_controller.rb]: Updated to `204` status code.
